include/pmacc/simulationControl/SimulationHelper.hpp

```diff
@@ -401,41 +404,97 @@ namespace pmacc
+        /** MPI request to find largest time step in the simulation */
+        MPI_Request signalMPI = MPI_REQUEST_NULL;
+        bool signalCreateCheckpoint = false;
+        bool signalStopSimulation = false;
+
         void checkSignals(uint32_t const currentStep)
         {
-            if(signal::received())
+            /* Avoid signal handling if the last signal is still processed.
+             * Signal handling in the first step is always allowed.
+             */
+            bool const handleSignals = handleSignalAtStep < currentStep || currentStep == 0u;
+            if(handleSignals && signal::received())
             {
+                /* Signals will not trigger actions directly, wait until handleSignalAtStep before
+                 * a signal is translated into an explicit action. This is required to avoid dead locks
+                 * with blocking collective operations. Each MPI rank can be in different time steps and phases
+                 * of the simulation, therefore we can not assume that all MPI ranks received the signal in the same
+                 * time step
+                 */
+
                 if(output)
                     std::cout << "SIGNAL: received." << std::endl;
 
-                // avoid deadlocks with collective MPI calls
-                Environment<>::get().Manager().waitForAllTasks();
                 // wait for possible more signals
                 std::this_thread::sleep_for(std::chrono::milliseconds(1000u));
 
-                uint32_t maxTimestep = 0u;
+                /* After a signal is received we need to perform one more time step to avoid dead-locks if a
+                 * simulation phase is using blocking MPI collectives. After the additional step we know that
+                 * all MPI ranks participated in MPI_Iallreduce.
+                 */
+                handleSignalAtStep = currentStep + 1;
 
                 // find largest time step of all MPI ranks
-                MPI_CHECK(MPI_Allreduce(
-                    &currentStep,
-                    &maxTimestep,
+                MPI_CHECK(MPI_Iallreduce(
+                    &handleSignalAtStep,
+                    &signalMaxTimestep,
                     1,
                     MPI_UINT32_T,
                     MPI_MAX,
-                    Environment<DIM>::get().GridController().getCommunicator().getMPIComm()));
+                    Environment<DIM>::get().GridController().getCommunicator().getMPISignalComm(),
+                    &signalMPI));
 
                 if(signal::createCheckpoint())
                 {
                     if(output)
-                        std::cout << "SIGNAL: active checkpointing for step " << maxTimestep << std::endl;
-                    // add a new checkpoint
-                    seqCheckpointPeriod.push_back(pluginSystem::TimeSlice(maxTimestep, maxTimestep));
+                        std::cout << "SIGNAL: Received at step " << currentStep << ". Schedule checkpointing. "
```

Comment:
Does it intentionally print to `cout` and not to the `log`?

Reply:
Yes, all other output in this class is printing to `std::cout` too. For pmacc we have no useful default log level, this is something I would like to change with a logger refactoring.